### PR TITLE
Add basic hardening guide to admin docs

### PR DESCRIPTION
**Pull request by KnightWhoSaysNeeeowWumPing**
Taken from my own comment on https://github.com/matrix-org/synapse/issues/5141#issuecomment-1013240289.

@H-Shay asked to add it to the documentation.

### Pull Request Checklist

* [x] Pull request is based on the develop branch
* [x] Pull request includes a [changelog file](https://matrix-org.github.io/synapse/latest/development/contributing_guide.html#changelog). The entry should:
  - Be a short description of your change which makes sense to users. "Fixed a bug that prevented receiving messages from other servers." instead of "Moved X method from `EventStore` to `EventWorkerStore`.".
  - Use markdown where necessary, mostly for `code blocks`.
  - End with either a period (.) or an exclamation mark (!).
  - Start with a capital letter.
  - Feel free to credit yourself, by adding a sentence "Contributed by @github_username." or "Contributed by [Your Name]." to the end of the entry.
* [x] Pull request includes a [sign off](https://matrix-org.github.io/synapse/latest/development/contributing_guide.html#sign-off)
* [x] [Code style](https://matrix-org.github.io/synapse/latest/code_style.html) is correct
  (run the [linters](https://matrix-org.github.io/synapse/latest/development/contributing_guide.html#run-the-linters))

Signed-off-by: KnightWhoSaysNeeeowWumPing KnightWhoSaysNeeeowWumPing@protonmail.com

### CODE REVIEW

**Review by anoadragon453 — CHANGES_REQUESTED** [review]
Thank you for working on this! Some suggestions are below, though note that this is still blocked on getting the sign-off above.

File: docs/usage/administration/hardening.md
Comment (richvdh) [reply docs/usage/administration/hardening.md:11]: this isn't adding an outbound rule, it's adding an inbound rule.

File: docs/usage/administration/hardening.md
Comment (richvdh) [inline docs/usage/administration/hardening.md:10]: there's no way to know for certain which ports a remote homeserver wants to listen on. Port 443 and 8448 are *common*, but for example `matrix.org` used 8443 until relatively recently.

In short, I think you'll need to make this rather broader.

File: docs/usage/administration/hardening.md
Comment (richvdh) [inline docs/usage/administration/hardening.md:11]: > although there seems to be some uncertainty, whether this kind of setup is recommended, see https://github.com/matrix-org/synapse/issues/2438)

it is absolutely recommended to use a reverse proxy for your federation port, whether it be :8448 or :443. If I were to suggest one thing for "hardening" synapse, putting it behind a reverse proxy would be it.

#2438 is quite old.

**Review by richvdh — CHANGES_REQUESTED** [review]

File: docs/usage/administration/hardening.md
Comment (KnightWhoSaysNeeeowWumPing) [reply docs/usage/administration/hardening.md:11]: Sorry, my bad - will change this typo..

File: docs/usage/administration/hardening.md
Comment (KnightWhoSaysNeeeowWumPing) [reply docs/usage/administration/hardening.md:21]: Thanks for pointing that out!

File: docs/usage/administration/hardening.md
Comment (KnightWhoSaysNeeeowWumPing) [reply docs/usage/administration/hardening.md:6]: I like your idea of a table, I'd just merge the `iptable` examples into their own column of said table for easier assignment..

File: docs/usage/administration/hardening.md
Comment (KnightWhoSaysNeeeowWumPing) [reply docs/usage/administration/hardening.md:10]: I think this boils down to the old security vs. functionality problem. In my case of a little friends-and-family homeserver, I had no issues (after adding the 8448 rule) yet. As soon as I encounter a connectivity problem with another homeserver, I review my firewall configuration and enable temporary logging of discarded packages. This way I am able to estimate the need for a new rule regarding a specific port.
But I understand that this is only my preferred way of doing things.. one could easily argue that less restrictive rules on the outbound chain impose only a minor security risk.
Regarding the guide, I'll try to make it clear that this is a more secure, less fail safe approach.

File: docs/usage/administration/hardening.md
Comment (KnightWhoSaysNeeeowWumPing) [reply docs/usage/administration/hardening.md:11]: I added a sentence for using a reverse proxy and removed the old 'uncertainty' part.

**Comment by KnightWhoSaysNeeeowWumPing** [comment]
> if you prefer sending your name privately instead, you can do so by sending a quick email to dco@matrix.org.

Thanks for reminding me.. I did just that!

> I'm assuming that KnightWhoSaysNeeeowWumPing is not legally identifiable

How can you tell? :smile:

File: docs/usage/administration/hardening.md
Comment (richvdh) [reply docs/usage/administration/hardening.md]: I'm afraid if your server does this, it is not compliant with the matrix spec. It might work most of the time for you, but please don't recommend it for others.

In general, I don't really think you get much security by restricting outbound connections by port.

Comment (KnightWhoSaysNeeeowWumPing) [reply docs/usage/administration/hardening.md]: So the general recommendation should not include any outbound restrictions, got you.
I'll update the guide accordingly!

**Comment by anoadragon453** [comment]
@KnightWhoSaysNeeeowWumPing the private sign-off process has now completed. Thank you for bearing with us on that; it's still relatively new.

And as a reminder, you won't need to do the same for future contributions - it only needs to be done once.

**Comment by richvdh** [comment]
Thank you for your work on this @KnightWhoSaysNeeeowWumPing!

As a team, we're not really sure how to proceed here. Concerns voiced when I mentioned it to the team earlier include:
* Whilst the content is largely correct, the styling, grammar and wording don't really have the level of polish we try to achieve with our documentation, and the inconsistency is jarring.
* It's a document the Synapse core team will need to take on and maintain. We aren't really in a position to advise how best to lock down your nginx, for example.
* The main thing #5141 is looking for is a page detailing *Synapse* security best practices - how to configure ports for Synapse, not giving worldwide access to replication or Admin API endpoints, as well as some default options one can change to reduce attack surface. Stuff outside of that should probably be limited to a footnote.
* For example: there's a lot of detail about nginx. It might be better to link to appropriate resources than to try to copy everything into our documentation. That said, I'm not aware of any great resources which I'd be particularly happy to link to from the official synapse documentation.

On the flip side, certainly there is useful information here, and it could give a useful basis to build from.

I'll discuss it more with the team.

**Comment by callahad** [comment]
Hi @KnightWhoSaysNeeeowWumPing, thank you so much for submitting this and jumping through the sign-off hoops.  This is a valuable contribution to our documentation. Unfortunately, we don't think it's *quite* the right fit for Synapse's main docs, but we're struggling to articulate exactly *why* that's the case.

As Rich touched on above, the Nginx-specific information is great, but we're worried about creating an obligation on the Synapse team itself to maintain those very specific instructions not just for Nginx, but also for many other common servers (Apache, Caddy, etc). This quickly becomes unwieldy, as seen in our installation docs.

So we know we want to keep new docs as high-level and universal as possible... but that means we're missing a place for more specific instructions, maintained by the community. Perhaps we should revive the wiki for those?

We're not asking you to do anything at this moment (we don't want to send you around in circles when we ourselves are indecisive).

If you don't object, I suspect the best outcome would be someone from the Synapse team to rewrite this page to more closely fit our subjective and ill-defined vision for Synapse's docs, directly crediting you in the process. In parallel, we're discussing bringing back the wiki to hold the more server-specific information you've provided in this pull request.

**Comment by KnightWhoSaysNeeeowWumPing** [comment]
> If you don't object, I suspect the best outcome would be someone from the Synapse team to rewrite this page to more closely fit our subjective and ill-defined vision for Synapse's docs, directly crediting you in the process. In parallel, we're discussing bringing back the wiki to hold the more server-specific information you've provided in this pull request.

Fine with me - after all, I only opened this as a PR because I was asked to do so. :upside_down_face: 
If someone ever searches for this type of information, it's out there - I get the caveats to include it into the docs as is.

**Pull request closed without merging**
